### DISCUSSION TITLE
Something like `x-for` but for repeating contents of an attribute?

### DISCUSSION THREAD

**Discussion by trusktr** [alpinejs/alpine#3733] [3. Feature ideas]
Hello! I'm working on [Lume](https://lume.io/), and with Lume elements we can do this:

```html
<lume-instanced-mesh
  has="box-geometry"
  colors="
    deeppink
    cyan
    yellow
  "
  positions="
    1 2 3
    4 5 6
    7 8 9
  "
></lume-instanced-mesh>
```

where this will render three boxes on the screen, in three different colors, and at three different positions in 3D space (x y z).

It would be useful to be able to repeat the content of an attribute. I didn't see a way in the docs.

How might a new feature look like? Kinda tricky to make this one ergonomic.

```html
<lume-instanced-mesh
  x-attr-positions-for="location in locations" positions="location.position +' '"
  x-attr-colors-for="location in locations" colors="location.color +' '"
></lume-instanced-mesh>
```

or

```html
<lume-instanced-mesh>
  <x-repeat-attr attr="positions" for="location in locations" value="location.position + ' '"></x-repeat-attr>
  <x-repeat-attr attr="colors" for="location in locations" value="location.color + ' '"></x-repeat-attr>
</lume-instanced-mesh>
```

**Comment by ekwoka** [comment]
Would probably be smart to put them into actual data, but assuming they HAVE to be in the attributes

```js
<template x-for="box in [$el.getAttribute('colors').trim().split('\n'), $el.getAttribute('positions').trim().split('\n')].reduce((a,b) a.flatMap(color => b.map(pos => ({ color, pos })))">
  <div x-text="box"></div>
</template>
```

I believe you could also just directly use `:colors` and then use Alpine.bound to get the colors, but it won't really make the code much shorter realistically.

**Reply by trusktr** [reply]
I'm not sure what you mean. What I'm describing is this:

Suppose that `locations` is an global variable containing an array. Now I want to write HTML that iterates that array to produce the content of a `positions` attribute on some element. I don't want to read _from_ the attribute on an element.

```html
<script>
  var locations = [...]
</script>

<some-element some-attribute="... fill this attribute with values based on that locations array ..."></some-element>
```

**Comment by SimoTod** [comment]
I would use a getter https://codepen.io/SimoTod/pen/mdaVNMp

**Reply by ekwoka** [reply]
Yes, definitely put it in the data.

The question is strangely presented, so I can't tell if needing the values in the actual dom attributes was important or not.

**Reply by trusktr** [reply]
Ah, I see, you both are saying to use `Alpine.data()`, in a way kinda like `new Vue()`. It wasn't clear from the other comment @ekwoka (I'm a total Alpine newb).

So this requires `x-data` for sure. I thought maybe Alpine would just read global variables.

What I'm trying to avoid is adding more JS, so that my new-to-HTML-non-JavaScript-knowing designer friend can start to describe some things in the simplest way possible.

**Reply by trusktr** [reply]
I was hoping something like this would simply work (and then something similar for attribute content):

```html
<script>
  var colors = ['red', 'green', 'blue']
</script>

<template x-for="color in colors">
  <div x-text="color"></div>
</template>

<script defer src="alpine.js"></script>
```

Nice and simple.

**Reply by trusktr** [reply]
Oh! This works:

```html
<script>
  var colors = ['red', 'green', 'blue']
</script>

<template x-data x-for="color in colors">
  <div x-text="color"></div>
</template>

<script defer src="alpine.js"></script>
```

(added empty `x-data` and it started working!)

**Reply by trusktr** [reply]
Aha!!!!!! And this works too!!

```html
<script>
  var colors = ['red', 'green', 'blue']
</script>

<template x-data x-for="color in colors">
  <div x-text="color" :some-attr="colors"></div>
</template>

<script defer src="alpine.js"></script>
```

The `some-attr` attributes will contain `"red,green,blue"`! Thanks for the hints!

**Reply by ekwoka** [reply]
Yup, just use `const`instead :)

**Reply by trusktr** [reply]
Not `using`? 😄 

What if I want to change the value later? Seems like `Alpine.data` is needed for that, for reactivity, while what I did is only good for initial values.

**Reply by ekwoka** [reply]
Yes, or Alpine.reactive if you want to be more custom about it. But you should probably use data.
